### ISSUE TITLE
Crash in QtSharedPointer::CustomDeleter in some cases

### ISSUE THREAD

**Issue by yan12125** [keepassxreboot/keepassxc#4877]
## Overview
When I was testing #4848, I found that in some cases KeePassXC crashes at exit. See Steps to Reproduce for details.

I use "Release Preview Bug report" as I don't get a crash with the same steps on 2.5.4.

## Steps to Reproduce
(Steps 1 and 2 are not needed if there are no recent databases)
1. Database -> Recent Databases -> Clear History
2. Restart KeePassXC
3. Tools -> Settings
4. Change nothing and hit Cancel
5. Hit Ctrl+Q to leave KeePassXC

## Expected Behavior
KeePassXC exits cleanly

## Actual Behavior
Segmentation fault. Backtrace from gdb:
```
#0  0x0000555555618f3f in QtSharedPointer::CustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::execute() (this=0x555556901b20)
    at /usr/include/qt/QtCore/qsharedpointer_impl.h:186
#1  0x0000555555618ced in QtSharedPointer::ExternalRefCountWithCustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::deleter(QtSharedPointer::ExternalRefCountData*) (self=0x555556901b10) at /usr/include/qt/QtCore/qsharedpointer_impl.h:204                                              
#2  0x00005555555db6f5 in QtSharedPointer::ExternalRefCountData::destroy() (this=0x555556901b10)
    at /usr/include/qt/QtCore/qsharedpointer_impl.h:148
#3  0x0000555555618dee in QSharedPointer<ISettingsPage>::deref(QtSharedPointer::ExternalRefCountData*) (dd=0x555556901b10)
    at /usr/include/qt/QtCore/qsharedpointer_impl.h:456
#4  0x0000555555618a44 in QSharedPointer<ISettingsPage>::deref() (this=0x555556901b30) at /usr/include/qt/QtCore/qsharedpointer_impl.h:451
#5  0x00005555556186ec in QSharedPointer<ISettingsPage>::~QSharedPointer() (this=0x555556901b30, __in_chrg=<optimized out>)
    at /usr/include/qt/QtCore/qsharedpointer_impl.h:309
#6  0x00005555556185fa in ApplicationSettingsWidget::ExtraPage::~ExtraPage() (this=0x555556901b30, __in_chrg=<optimized out>)
    at /build/keepassxc-git/src/keepassxc/src/gui/ApplicationSettingsWidget.cpp:35
#7  0x0000555555618e5c in QList<ApplicationSettingsWidget::ExtraPage>::node_destruct(QList<ApplicationSettingsWidget::ExtraPage>::Node*, QList<ApplicationSettingsWidget::ExtraPage>::Node*) (this=0x55555618c738, from=0x55555687cb10, to=0x55555687cb28) at /usr/include/qt/QtCore/qlist.h:524    
#8  0x0000555555618af3 in QList<ApplicationSettingsWidget::ExtraPage>::dealloc(QListData::Data*) (this=0x55555618c738, data=0x55555687cb00)
    at /usr/include/qt/QtCore/qlist.h:921
#9  0x00005555556187dc in QList<ApplicationSettingsWidget::ExtraPage>::~QList() (this=0x55555618c738, __in_chrg=<optimized out>)
    at /usr/include/qt/QtCore/qlist.h:874
#10 0x000055555560c846 in ApplicationSettingsWidget::~ApplicationSettingsWidget() (this=0x55555618c6d0, __in_chrg=<optimized out>)
    at /build/keepassxc-git/src/keepassxc/src/gui/ApplicationSettingsWidget.cpp:157
#11 0x000055555560c894 in ApplicationSettingsWidget::~ApplicationSettingsWidget() (this=0x55555618c6d0, __in_chrg=<optimized out>)
    at /build/keepassxc-git/src/keepassxc/src/gui/ApplicationSettingsWidget.cpp:159
#12 0x00007ffff6c8ac3e in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#13 0x00007ffff78271de in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#14 0x00007ffff782755a in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#15 0x00007ffff6c8ac3e in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#16 0x00007ffff78271de in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#17 0x00007ffff799b06a in QStackedWidget::~QStackedWidget() () at /usr/lib/libQt5Widgets.so.5
#18 0x00007ffff6c8ac3e in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#19 0x00007ffff78271de in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#20 0x00007ffff782755a in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#21 0x00007ffff6c8ac3e in QObjectPrivate::deleteChildren() () at /usr/lib/libQt5Core.so.5
#22 0x00007ffff78271de in QWidget::~QWidget() () at /usr/lib/libQt5Widgets.so.5
#23 0x00005555555ebe8e in MainWindow::~MainWindow() (this=0x7fffffffdd00, __in_chrg=<optimized out>)
    at /build/keepassxc-git/src/keepassxc/src/gui/MainWindow.cpp:574
#24 0x00005555555c58e5 in main(int, char**) (argc=1, argv=0x7fffffffdf78) at /build/keepassxc-git/src/keepassxc/src/main.cpp:132
```

## Context

KeePassXC - Version 2.6.0-snapshot
Build Type: Snapshot
Revision: c68f487

(actually c68f487 with patch #4848)

Qt 5.15.0
Debugging mode is enabled.

Operating system: Arch Linux
CPU architecture: x86_64
Kernel: linux 5.7.2-arch1-1

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- Secret Service Integration

Cryptographic libraries:
 libgcrypt 1.8.5

Operating System: Arch Linux
Desktop Env: LXQt
Windowing System: X11

I use PKGBUILD modified from the AUR one to integrate patch #4848 and enable debugging. Here is the full PKGBUILD.
```
# Maintainer: Daniel Landau <daniel@landau.fi>
# Contributor: Maxqia <contrib@maxqia.com>

# The following people have contributed to keepassx-git package
# Contributor: Lev Lybin <aur@devtrue.net>
# Contributor: Jamie Macdonald <jamie.alban@gmail.com>
# Contributor: Alucryd <alucryd at gmail dot com>
# Contributor: Paolo Stivanin <admin at polslinux dot it>

pkgname=keepassxc-git
_gitname=keepassxc
pkgver=2.6.0.beta1.r3.gc68f4872
pkgrel=1
pkgdesc="Cross-platform community-driven port of KeePass Password Safe"
arch=('i686' 'x86_64')
url="https://github.com/keepassxreboot/keepassxc"
license=('GPL2')
depends=('qt5-base' 'qt5-x11extras' 'qt5-translations' 'qt5-imageformats'
         'libxtst' 'shared-mime-info' 'hicolor-icon-theme' 'desktop-file-utils'
         'yubikey-personalization' 'qt5-svg' 'qrencode' 'quazip'
         'libsodium' 'libgcrypt' 'argon2' 'zlib')
#install=keepassxc.install
makedepends=('git' 'intltool' 'cmake' 'qt5-tools')
provides=(org.freedesktop.secrets)
conflicts=('keepassxc')
replaces=('keepassx-http' 'keepassx-reboot-git')
options=(!emptydirs !buildflags !strip)
source=(git+https://github.com/keepassxreboot/keepassxc.git#branch=release/2.6.0
        keepassxc-issue4835.patch::https://github.com/keepassxreboot/keepassxc/pull/4848.patch)
md5sums=('SKIP'
         'f2ec0878cd2b58a5e838752a116fd6a0')

pkgver() {
    cd "${_gitname}"
    git describe --long | sed 's/^FOO-//;s/\([^-]*-g\)/r\1/;s/-/./g'
}

prepare() {
    cd "${_gitname}"
    patch -Np1 -i ../keepassxc-issue4835.patch
    mkdir -p build
    sed -i'' -e 's:/usr/local/share/man/:${CMAKE_INSTALL_MANDIR}/:' src/cli/CMakeLists.txt
}

build() {
    cd "${_gitname}/build"
    export CXXFLAGS="$CXXFLAGS -Wno-unused-result"
    cmake -DCMAKE_INSTALL_PREFIX=/usr \
        -DCMAKE_INSTALL_BINDIR=/usr/bin \
        -DCMAKE_INSTALL_LIBDIR=/usr/lib \
        -DCMAKE_VERBOSE_MAKEFILE=OFF \
        -DWITH_GUI_TESTS=ON \
        -DWITH_XC_ALL=ON \
        -DWITH_XC_KEESHARE_SECURE=ON \
        -DCMAKE_BUILD_TYPE=Debug ..
    make VERBOSE=1
}

#check() {
#    cd "${_gitname}/build"
#    make test
#}

package() {
    cd "${_gitname}/build"
    make DESTDIR="${pkgdir}" install
}
```

[1] https://aur.archlinux.org/packages/keepassxc-git

**Comment by yan12125** [comment]
Seems a double deletion bug. Relevant lines from valgrind using the same steps to reproduce as mentioned above:
```
==1472== Invalid read of size 8
==1472==    at 0x1CCF38: QtSharedPointer::CustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::execute() (qsharedpointer_impl.h:186)
==1472==    by 0x1CCCEC: QtSharedPointer::ExternalRefCountWithCustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::deleter(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:204)
==1472==    by 0x18F6F4: QtSharedPointer::ExternalRefCountData::destroy() (qsharedpointer_impl.h:148)
==1472==    by 0x1CCDED: QSharedPointer<ISettingsPage>::deref(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:456)
==1472==    by 0x1CCA43: QSharedPointer<ISettingsPage>::deref() (qsharedpointer_impl.h:451)
==1472==    by 0x1CC6EB: QSharedPointer<ISettingsPage>::~QSharedPointer() (qsharedpointer_impl.h:309)
==1472==    by 0x1CC5F9: ApplicationSettingsWidget::ExtraPage::~ExtraPage() (ApplicationSettingsWidget.cpp:35)
==1472==    by 0x1CCE5B: QList<ApplicationSettingsWidget::ExtraPage>::node_destruct(QList<ApplicationSettingsWidget::ExtraPage>::Node*, QList<ApplicationSettingsWidget::ExtraPage>::Node*) (qlist.h:524)
==1472==    by 0x1CCAF2: QList<ApplicationSettingsWidget::ExtraPage>::dealloc(QListData::Data*) (qlist.h:921)
==1472==    by 0x1CC7DB: QList<ApplicationSettingsWidget::ExtraPage>::~QList() (qlist.h:874)
==1472==    by 0x1C0845: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:157)
==1472==    by 0x1C0893: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:159)
==1472==  Address 0xdea4210 is 16 bytes inside a block of size 48 free'd
==1472==    at 0x483B08B: operator delete(void*, unsigned long) (vg_replace_malloc.c:593)
==1472==    by 0x39363E: FdoSecretsPlugin::~FdoSecretsPlugin() (FdoSecretsPlugin.h:39)
==1472==    by 0x5BE4C3D: QObjectPrivate::deleteChildren() (in /usr/lib/libQt5Core.so.5.15.0)
==1472==    by 0x4C7C1DD: QWidget::~QWidget() (in /usr/lib/libQt5Widgets.so.5.15.0)
==1472==    by 0x293EFF: DatabaseTabWidget::~DatabaseTabWidget() (DatabaseTabWidget.cpp:75)
==1472==    by 0x293F21: DatabaseTabWidget::~DatabaseTabWidget() (DatabaseTabWidget.cpp:77)
==1472==    by 0x5BE4C3D: QObjectPrivate::deleteChildren() (in /usr/lib/libQt5Core.so.5.15.0)
==1472==    by 0x4C7C1DD: QWidget::~QWidget() (in /usr/lib/libQt5Widgets.so.5.15.0)
==1472==    by 0x4C7C559: QWidget::~QWidget() (in /usr/lib/libQt5Widgets.so.5.15.0)
==1472==    by 0x5BE4C3D: QObjectPrivate::deleteChildren() (in /usr/lib/libQt5Core.so.5.15.0)
==1472==    by 0x4C7C1DD: QWidget::~QWidget() (in /usr/lib/libQt5Widgets.so.5.15.0)
==1472==    by 0x4DF0069: QStackedWidget::~QStackedWidget() (in /usr/lib/libQt5Widgets.so.5.15.0)
==1472==  Block was alloc'd at
==1472==    at 0x4839DEF: operator new(unsigned long) (vg_replace_malloc.c:342)
==1472==    by 0x199DC5: MainWindow::MainWindow() (MainWindow.cpp:179)
==1472==    by 0x1794B7: main (main.cpp:132)
==1472== 
==1472== Jump to the invalid address stated on the next line
==1472==    at 0x0: ???
==1472==    by 0x1CCCEC: QtSharedPointer::ExternalRefCountWithCustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::deleter(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:204)
==1472==    by 0x18F6F4: QtSharedPointer::ExternalRefCountData::destroy() (qsharedpointer_impl.h:148)
==1472==    by 0x1CCDED: QSharedPointer<ISettingsPage>::deref(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:456)
==1472==    by 0x1CCA43: QSharedPointer<ISettingsPage>::deref() (qsharedpointer_impl.h:451)
==1472==    by 0x1CC6EB: QSharedPointer<ISettingsPage>::~QSharedPointer() (qsharedpointer_impl.h:309)
==1472==    by 0x1CC5F9: ApplicationSettingsWidget::ExtraPage::~ExtraPage() (ApplicationSettingsWidget.cpp:35)
==1472==    by 0x1CCE5B: QList<ApplicationSettingsWidget::ExtraPage>::node_destruct(QList<ApplicationSettingsWidget::ExtraPage>::Node*, QList<ApplicationSettingsWidget::ExtraPage>::Node*) (qlist.h:524)
==1472==    by 0x1CCAF2: QList<ApplicationSettingsWidget::ExtraPage>::dealloc(QListData::Data*) (qlist.h:921)
==1472==    by 0x1CC7DB: QList<ApplicationSettingsWidget::ExtraPage>::~QList() (qlist.h:874)
==1472==    by 0x1C0845: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:157)
==1472==    by 0x1C0893: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:159)
==1472==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
==1472== 
==1472== 
==1472== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==1472==  Bad permissions for mapped region at address 0x0
==1472==    at 0x0: ???
==1472==    by 0x1CCCEC: QtSharedPointer::ExternalRefCountWithCustomDeleter<ISettingsPage, QtSharedPointer::NormalDeleter>::deleter(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:204)
==1472==    by 0x18F6F4: QtSharedPointer::ExternalRefCountData::destroy() (qsharedpointer_impl.h:148)
==1472==    by 0x1CCDED: QSharedPointer<ISettingsPage>::deref(QtSharedPointer::ExternalRefCountData*) (qsharedpointer_impl.h:456)
==1472==    by 0x1CCA43: QSharedPointer<ISettingsPage>::deref() (qsharedpointer_impl.h:451)
==1472==    by 0x1CC6EB: QSharedPointer<ISettingsPage>::~QSharedPointer() (qsharedpointer_impl.h:309)
==1472==    by 0x1CC5F9: ApplicationSettingsWidget::ExtraPage::~ExtraPage() (ApplicationSettingsWidget.cpp:35)
==1472==    by 0x1CCE5B: QList<ApplicationSettingsWidget::ExtraPage>::node_destruct(QList<ApplicationSettingsWidget::ExtraPage>::Node*, QList<ApplicationSettingsWidget::ExtraPage>::Node*) (qlist.h:524)
==1472==    by 0x1CCAF2: QList<ApplicationSettingsWidget::ExtraPage>::dealloc(QListData::Data*) (qlist.h:921)
==1472==    by 0x1CC7DB: QList<ApplicationSettingsWidget::ExtraPage>::~QList() (qlist.h:874)
==1472==    by 0x1C0845: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:157)
==1472==    by 0x1C0893: ApplicationSettingsWidget::~ApplicationSettingsWidget() (ApplicationSettingsWidget.cpp:159)
```

And the crash is gone if I build KeePassXC without the FDO secrets plugin.

**Comment by yan12125** [comment]
The issue itself is fixed, while my pull request needs yet another fix. Will follow-up there.

**Comment by droidmonkey** [comment]
I cannot replicate this at all under Ubuntu 20.04 using the sequence defined above.

**Comment by yan12125** [comment]
Maybe Qt recently changed how a shared pointer is destructed. Looks like Ubuntu 20.04 has Qt 5.12 [1] while I use Qt 5.15.

And, FWIW, your new fix (#4884) also works find. Thanks!

[1] https://packages.ubuntu.com/source/focal/qtbase-opensource-src
